`valid_cur_vcpu`

michaelmcinerney opened the pull request:
This adds the new invariant `valid_cur_vcpu` that says that the active current vcpu is the vcpu of the current thread.

michaelmcinerney commented:
I stupidly imported a file generically that doesn't exist on arches other than ARM_HYP. But seeing that this PR is targetting the spec/proof update, I might wait until that PR is merged before significantly updating this. Otherwise, the AInvs content for ARM_HYP should still be good to review. Happy to mark this PR as a draft for now if that's preferable

lsf37 commented:
Since the tests won't be working until this is sorted (so we won't merge accidentally), I think it's fine to leave it on non-draft for now. For reviewing it is good enough.

michaelmcinerney commented on proof/invariant-abstract/ARM_HYP/ArchVCPU_AI.thy:
```diff
@@ -0,0 +1,729 @@
+(*
+ * Copyright 2022, ???
```
Still need to add the proper copyright. The only thing I took from Sidney and Mitch's work was the definition of `active_cur_vcpu_of`. Should this be only UNSW then?

lsf37 replied:
Yes, that is good.

lsf37 commented on proof/invariant-abstract/ARM_HYP/ArchVCPU_AI.thy:
```diff
@@ -0,0 +1,727 @@
+
+end
+
+end
```
please add line feed at last line

lsf37 commented on proof/invariant-abstract/ARM_HYP/ArchVCPU_AI.thy:
```diff
@@ -0,0 +1,727 @@
+  and active_cur_vcpu_of: "\<And>P. \<lbrace>\<lambda>s. P (active_cur_vcpu_of s) \<and> Q s\<rbrace>
+                                f \<lbrace>\<lambda>_ s. P (active_cur_vcpu_of s)\<rbrace>"
+  shows "\<lbrace>\<lambda>s. valid_cur_vcpu s \<and> Q s\<rbrace> f \<lbrace>\<lambda>_. valid_cur_vcpu\<rbrace>"
+  apply (clarsimp simp: valid_cur_vcpu_def valid_def)
```
could this `clarsimp` be an `unfolding ..`?

lsf37 commented on proof/invariant-abstract/ARM_HYP/ArchVCPU_AI.thy:
```diff
@@ -0,0 +1,727 @@
+  "\<lbrace>arch_tcb_at (\<lambda>itcb. itcb_vcpu itcb = v) t\<rbrace>
+   vcpu_switch v
+   \<lbrace>\<lambda>_ s. valid_cur_vcpu (s\<lparr>cur_thread := t\<rparr>)\<rbrace>"
+  apply (clarsimp simp: vcpu_switch_def)
```
this one also looks like it could be an `unfolding` (it's not super important, but it's nice when the unfolding step is stable and you know in maintenance that nothing interesting is supposed to be happening here).

corlewis replied:
This feels like a specific enough guideline that it would be worth recording. Would [`docs/style`](https://github.com/seL4/l4v/blob/master/docs/Style.thy#L251) be the right place to put it?

lsf37 replied:
Yes, that would be a good place

lsf37 commented on proof/invariant-abstract/ARM_HYP/ArchVCPU_AI.thy:
```diff
@@ -0,0 +1,727 @@
+  for active_cur_vcpu_of[wp]: "\<lambda>s. P (active_cur_vcpu_of s)"
+  (wp: crunch_wps simp: active_cur_vcpu_of_def)
+
+lemma find_free_hw_asid_active_cur_vcpu_of[wp]:
+  "find_free_hw_asid \<lbrace>\<lambda>s. P (active_cur_vcpu_of s)\<rbrace>"
+  apply (clarsimp simp: find_free_hw_asid_def find_pd_for_asid_assert_def)
+  apply (intro hoare_seq_ext[OF _ gets_inv])
+  apply (clarsimp split: option.splits)
+  apply (rule hoare_seq_ext_skip, wpsimp)+
+   apply (clarsimp simp: active_cur_vcpu_of_def pred_tcb_at_def obj_at_def valid_cur_vcpu_def)
+  apply (wpsimp simp: find_pd_for_asid_assert_def)
+  done
```
```suggestion
lemma active_cur_vcpu_of_next_asid[simp]:
  "active_cur_vcpu_of (s⦇arch_state := arch_state s⦇arm_next_asid := v⦈⦈) = active_cur_vcpu_of s"
  by (clarsimp simp: active_cur_vcpu_of_def pred_tcb_at_def obj_at_def valid_cur_vcpu_def)

lemma find_free_hw_asid_active_cur_vcpu_of[wp]:
  "find_free_hw_asid ⦃λs. P (active_cur_vcpu_of s)⦄"
  by (wpsimp simp: find_free_hw_asid_def find_pd_for_asid_assert_def)
```

lsf37 commented on proof/invariant-abstract/ARM_HYP/ArchVCPU_AI.thy:
```diff
@@ -0,0 +1,727 @@
+  "\<lbrace>\<lambda>s. valid_cur_vcpu s \<and> sym_refs (state_hyp_refs_of s)\<rbrace>
+   dissociate_vcpu_tcb vr t
+   \<lbrace>\<lambda>_. valid_cur_vcpu\<rbrace>"
+  apply (clarsimp simp: dissociate_vcpu_tcb_def)
+  apply (clarsimp simp: arch_thread_get_def)
+  apply (rule hoare_seq_ext[OF _ gets_the_sp])
+  apply (clarsimp simp: get_vcpu_def bind_assoc)
+  apply (rule hoare_seq_ext[OF _ get_object_sp])
+  apply (clarsimp split: kernel_object.splits arch_kernel_obj.splits)
+  apply (rule hoare_seq_ext[OF _ assert_sp])
+  apply (rule hoare_seq_ext[OF _ gets_sp])
+  apply (simp flip: bind_assoc)
+  apply (rule hoare_seq_ext)
+   apply wpsimp
+  apply (simp add: bind_assoc)
```
This looks like too much explicit forward single-stepping to me for a proof like this. Those single steps are very brittle in maintenance, because they spell out each single step of the function in the proof, so even small reorderings/additions will break. They shouldn't really be necessary here. Of course, that's hard to say from just the diff. Some of that could be because of the inline unfolding of `vcpu_invalidate_active`, which probably needs its own lemma(s).

The lemma below `associate_vcpu_tcb_valid_cur_vcpu` has the same flavour.`wp`/`wpsimp` should be effective on these, but they probably need some hand holding in the form of lemmas.

lsf37 commented on proof/invariant-abstract/ARM_HYP/ArchVCPU_AI.thy:
```diff
@@ -0,0 +1,727 @@
+lemma send_fault_ipc_arch_tcb_at[wp]:
+  "send_fault_ipc tptr fault \<lbrace>arch_tcb_at P t\<rbrace>"
+  unfolding send_fault_ipc_def thread_set_def Let_def
+  apply (rule validE_valid)
+  apply (rule hoare_seq_ext_skipE, solves wpsimp)+
+  apply (wpsimp wp: set_object_wp)
+  apply (clarsimp simp: pred_tcb_at_def obj_at_def get_tcb_def)
+  done
```
`wp` gets stuck at `lookup_cap` because we haven't provided any full `wp` rules for it, but we do have a rule for `arch_tcb_at` and all the implications are just accumulated information that is not actually necessary for that to be preserved. `hoare_drop_imps` will get rid of these implications automatically and `hoare_vcg_all_lift_R` will deal with the quantifiers that are introduced in the case distinction:
```suggestion
  by (wpsimp wp: set_object_wp hoare_drop_imps hoare_vcg_all_lift_R
             simp: pred_tcb_at_def obj_at_def get_tcb_def)
```

lsf37 commented on proof/invariant-abstract/ARM_HYP/ArchVCPU_AI.thy:
```diff
@@ -0,0 +1,727 @@
+  apply (case_tac fault_opt; clarsimp)
+   apply wpsimp
+  apply (rule hoare_seq_ext_skip, solves wpsimp)+
+  apply (wpsimp wp: set_object_wp hoare_drop_imps | intro conjI impI)+
```
The `intro conjI impI` is not a good pattern in combination with `wpsimp`, it's usually a sign that something has gone wrong. In this case it's splitting an `if` into a conjunction instead of progressing with `wp`.

Try 
`  apply (wpsimp wp: set_object_wp hoare_drop_imps split_del: split_if)`
instead (note no `+`, which `wpsimp` should never need for a single goal).

This specific thing (too eager `if_split`) is a problem with the `clarsimp` inside `wpsimp` being too eager. We could try to remove `if_split` in there by default, but we'd get lots of annoying warnings in contexts where `if_split` is already removed. Maybe we should put this one on the list of things to spend some time thinking about.

Xaphiosis replied:
Dropping the `if_split` is one way, and `hoare_drop_imps` is useful when you don't need to the condition of the `if`. If you *do* need the condition of the `if`, consider adding the simp rule `if_apply_def2` which converts `if` to implications, then you can decide whether you want to drop them or not. I don't know if this applies here, but it's a very useful tool to know about.

lsf37 replied:
The `if` is in the code this time (somewhere deep in a term), so this specific rule wouldn't help here, but in general it's a good idea to check, yes.

lsf37 commented on proof/invariant-abstract/ARM_HYP/ArchVCPU_AI.thy:
```diff
@@ -0,0 +1,727 @@
+lemma do_reply_transfer_arch_tcb_at[wp]:
+  "do_reply_transfer sender receiver slot grant \<lbrace>arch_tcb_at P t\<rbrace>"
+  unfolding do_reply_transfer_def thread_set_def
+  apply (rule hoare_seq_ext_skip, solves wpsimp, clarsimp?)+
```
We should stop using this pattern in a single ad-hoc proof style. I was fine with it as emergency measure for some of the RT proofs, but it's not a good thing to use everywhere. Unless the part we're splitting off is itself some big case distinction or something like that, this should never be necessary. `wpsimp` alone will solve all of these parts automatically if the harder later part is already solved.

If we really want to do this, then we should make it a proof method that is carefully constructed and deals natively with e.g. the `K_bind` that things get stuck on here.

lsf37 commented on proof/invariant-abstract/ARM_HYP/ArchVCPU_AI.thy:
```diff
@@ -0,0 +1,727 @@
+  (simp: reschedule_required_def set_scheduler_action_def tcb_sched_action_def set_tcb_queue_def
+         get_tcb_queue_def)
+
+lemma do_reply_transfer_arch_tcb_at[wp]:
+  "do_reply_transfer sender receiver slot grant \<lbrace>arch_tcb_at P t\<rbrace>"
```
So what this proof is getting stuck on is the unfolding (wp-rule use) of `set_object`. That is one level too low. In fact, unfolding `thread_set` is already one level too low, because `thread_set` with fault updates should always preserve `arch_tcb_at`. Once you have that, the lemma becomes almost entirely automatic:

```isabelle
lemma thread_set_fault_arch_tcb_at[wp]:
  "thread_set (tcb_fault_update f) r ⦃arch_tcb_at P t⦄"
  unfolding thread_set_def
  by (wpsimp wp: set_object_wp simp: pred_tcb_at_def obj_at_def get_tcb_def)

lemma do_reply_transfer_arch_tcb_at[wp]:
  "do_reply_transfer sender receiver slot grant ⦃arch_tcb_at P t⦄"
  unfolding do_reply_transfer_def
  by (wpsimp wp: gts_wp split_del: if_split)
```

lsf37 commented on proof/invariant-abstract/ARM_HYP/ArchVCPU_AI.thy:
```diff
@@ -0,0 +1,727 @@
+   defer
+   subgoal for tcb_ptr ntfn_ptr_opt
+     by (case_tac ntfn_ptr_opt; wpsimp)
+  \<comment> \<open>ThreadControl\<close>
+  apply (rename_tac target slot faultep mcp priority croot vroot buffer)
+  apply (rule validE_valid)
+  apply (rule hoare_seq_ext_skipE, solves wpsimp)+
+  apply (rule hoare_seq_ext_skipE)
+   apply (case_tac croot; clarsimp)
+    apply wpsimp
+   apply (rule_tac Q="\<lambda>_. ?pre" and E="\<lambda>_. ?pre" in hoare_post_impErr)
+     apply (rule hoare_seq_ext_skipE)
+      apply (wpsimp wp: cap_delete_valid_cur_vcpu)
+     apply (wpsimp wp: check_cap_inv)
+    apply simp
+   apply simp
+  apply (rule hoare_seq_ext_skipE)
+   apply (case_tac vroot; clarsimp)
+    apply wpsimp
+   apply (rule_tac Q="\<lambda>_. ?pre" and E="\<lambda>_. ?pre" in hoare_post_impErr)
+     apply (rule hoare_seq_ext_skipE)
+      apply (wpsimp wp: cap_delete_valid_cur_vcpu)
+     apply (wpsimp wp: check_cap_inv)
+    apply simp
+   apply simp
+  apply (rule hoare_seq_ext_skipE)
+   apply (case_tac buffer; clarsimp)
+    apply wpsimp
+   apply (rule hoare_seq_ext_skipE)
+    apply (wpsimp wp: cap_delete_valid_cur_vcpu)
+   apply (wpsimp wp: check_cap_inv hoare_drop_imps thread_set_hyp_refs_trivial thread_set_valid_cur_vcpu)
+  apply wpsimp
+  done
```
This is one where forward stepping with invariant make sense, because with pure backwards wp reasoning would blow up over the case distinctions. No need to do all those case distinctions manually, though, `wpsimp` can do them on its own. In fact, you can nicely collapse the entire proof into:

```suggestion
  \<comment> \<open>ThreadControl\<close>
  apply (rule validE_valid)
  apply (rule hoare_seq_ext_skipE, solves ‹wpsimp wp: check_cap_inv›)+
  by (wpsimp wp: check_cap_inv hoare_drop_imps thread_set_hyp_refs_trivial thread_set_valid_cur_vcpu)
```

lsf37 replied:
Even for this, though, we should ideally use a proof method that does the skipping step, reliably handles special cases, and works for all variants of `valid`, `validE`, etc.

Not sure it makes sense to put that into this PR, but it's not good to let the assembly-level proof steps proliferate too much, because they'll be hard to remove later if we wait too long.

lsf37 reviewed (approved):
Overall I'm happy with this, my current pet peeve is the forward stepping with `hoare_seq_ext_skip` :-) This is usually too manual and hard to maintain.

There is only one case that I can see where it really is beneficial, otherwise they are usually a sign of something having gone wrong in the proofs with `wp`, either a missing rule or unfolding too deeply which leads to the same effect (not hitting the spot where the automation works).

That one case where the forward stepping is better is when `wp` would otherwise lead to exponentially many cases. You get that when you have a chain of case distinctions -- the forward stepping joins the two case branches every time, so doesn't blow up. You can't do that easily with `wp`.

For those cases, and maybe for proof exploration, we should wrap up the forward stepping into a proof method. I'm fine with doing that separately, as long as we don't leave too long.

michaelmcinerney commented:
Thanks for the review, Gerwin. I've hopefully addressed all your comments.

I've included an Eisbach method that hopefully achieves what we would be after. I didn't want the method to apply `rule valid_validE` unless `rule validE_valid` could be applied, so that the method wouldn't mess with the proof state any more than necessary. If we aren't so worried about that, the method could be made slightly less repetitive:

```
method hoare_skip uses wp simp =
  (rule validE_valid)?,
  (rule hoare_seq_ext_skip | rule hoare_seq_ext_skipE),
  solves <wpsimp wp: wp simp: simp>,
  (rule valid_validE)?,
  clarsimp?
```

lsf37 commented:
> Thanks for the review, Gerwin. I've hopefully addressed all your comments.

:+1: Looks good!

> I've included an Eisbach method that hopefully achieves what we would be after. I didn't want the method to apply `rule valid_validE` unless `rule validE_valid` could be applied, so that the method wouldn't mess with the proof state any more than necessary. If we aren't so worried about that, the method could be made slightly less repetitive:
> 
> ```
> method hoare_skip uses wp simp =
>   (rule validE_valid)?,
>   (rule hoare_seq_ext_skip | rule hoare_seq_ext_skipE),
>   solves <wpsimp wp: wp simp: simp>,
>   (rule valid_validE)?,
>   clarsimp?
> ```

The proof already looks a lot nicer, but the method still needs a bit more refinement:

 - we shouldn't end in `clarsimp?`, because that can strongly affect the remaining goal (incl the monadic code) instead of just splitting things off. If it is for resolving `K_bind`, we should use/make a specific rule for it (`hoare_seq_ext_nobind` looks like the one to try before `hoare_seq_ext` to avoid that particular issue)

 - instead of the validE -> skip -> validE conversion, we should make rules that do this in one step, i.e. produce rules out of `hoare_seq_ext_nobind` that are already properly chained together so only combinations that make sense are possible. 

 - we should complete the rule set so it works for `validE_R` and `validR_R`

 - naming: the hardest part :-). "skip" is what we're used to, but it's not a good name. We're not really skipping anything.  Maybe `inv_step`? Doesn't quite express that it works from the front when everything else works from the back. Is `forward_inv_step` too long?

- taken all of the above together, we should probably make a named rule set for these. The order of rules in that set is important (e.g. so the ones with K_bind are tried first). With a rule set the method can then become just

       rule forward_inv_step_rules, solves <wpsimp wp: wp simp: simp>

   That would have the added advantage that if the `wpsimp` is not quite strong enough on its own, you can do a separate `apply (rule forward_inv_step_rules)` in the proof, and it should already take care of all the special cases, just leaving you with the two main wp goals to be solved (not to be abused, though :-))

michaelmcinerney commented:
Those are all very good points. I really like the idea of a named rule set. I'll look into that

michaelmcinerney commented:
I think I have a good rule set for the `forward_inv_step_rules` set. Where should the Eisbach method go? Immediately after the rules? I think that makes most sense. Or in the `Eisbach_Methods.thy` file?

lsf37 commented:
> I think I have a good rule set for the `forward_inv_step_rules` set. Where should the Eisbach method go? Immediately after the rules? I think that makes most sense. Or in the `Eisbach_Methods.thy` file?

The Eisbach_Methods file is for more general ones, this one would fit somewhere in the NondetMonadVCG space I think, close to where the rules are available.

lsf37 commented:
> > I think I have a good rule set for the `forward_inv_step_rules` set. Where should the Eisbach method go? Immediately after the rules? I think that makes most sense. Or in the `Eisbach_Methods.thy` file?
> 
> The Eisbach_Methods file is for more general ones, this one would fit somewhere in the NondetMonadVCG space I think, close to where the rules are available.

Thinking more about it, it might be worth putting those rules and the method into their own theory file, actually (and including that somewhere transparently so that we don't need to change other imports).

michaelmcinerney commented:
> Thinking more about it, it might be worth putting those rules and the method into their own theory file, actually (and including that somewhere transparently so that we don't need to change other imports).

If this new file imports `NonDetMonadVCG`, then I'm not sure how I would go about not changing other imports. `NonDetMonadVCG` is imported a dozen or so times, so I'm guessing we wouldn't want to replace all of those imports with importing this new file. Or maybe if I import the new file in `Include_AI`? Then it would at least be available everywhere in AInvs, Refine, CRefine

lsf37 commented:
> If this new file imports `NonDetMonadVCG`, then I'm not sure how I would go about not changing other imports. `NonDetMonadVCG` is imported a dozen or so times, so I'm guessing we wouldn't want to replace all of those imports with importing this new file. 

Yes, that'd be too much. But we could include the new file in `NonDetMonadVCG`, provided the lemmas that are needed in the new file can be moved into the new file without dragging too many other things with them.

michaelmcinerney commented:
> Yes, that'd be too much. But we could include the new file in `NonDetMonadVCG`, provided the lemmas that are needed in the new file can be moved into the new file without dragging too many other things with them.

The new rules for the new file require `hoare_seq_ext` and the `validE` and "nobind" versions, all of which are in `NonDetMonadVCG`. We wouldn't want to move those to the new file, would we?

lsf37 commented:
Cool, new version looks good to me!

corlewis commented on proof/invariant-abstract/ARM_HYP/ArchArch_AI.thy:
```diff
@@ -640,12 +640,11 @@ lemma max_index_upd_no_cap_to:
-    perform_asid_control_invocation aci
-  \<lbrace>\<lambda>y. st_tcb_at P t\<rbrace>"
+lemma perform_asid_control_invocation_pred_tcb_at:
+  "\<lbrace>\<lambda>s. pred_tcb_at proj Q t s \<and> st_tcb_at (P and (Not \<circ> inactive) and (Not \<circ> idle)) t s
+        \<and> ct_active s \<and> invs s \<and> valid_aci aci s\<rbrace>
+   perform_asid_control_invocation aci
+   \<lbrace>\<lambda>_. pred_tcb_at proj Q t\<rbrace>"
```
Apologies for commenting on this way too late, but I just noticed that the `P` in this precondition no longer seems necessary. Leaving it in like this shouldn't break anything but it might lead to confusing schematic variables when this rule is used

lsf37 replied:
Ah yes, you're right, I missed that. We should fix this in a new PR. I don't think it'll break much (or anything), but it's better to not have to worry about unused schematics popping up.

corlewis commented on proof/invariant-abstract/Untyped_AI.thy:
```diff
@@ -3868,26 +3868,33 @@ lemmas invoke_untyp_invs[wp] =
-         and ct_active and valid_untyped_inv ui\<rbrace>
-     invoke_untyped ui
-   \<lbrace>\<lambda>rv. \<lambda>s :: 'state_ext state. st_tcb_at P t s\<rbrace>"
+lemma invoke_untyped_pred_tcb_at:
+  "\<lbrace>\<lambda>s. pred_tcb_at proj Q t s \<and> invs s \<and> st_tcb_at (P and (Not \<circ> inactive) and (Not \<circ> idle)) t s
+        \<and> ct_active s \<and> valid_untyped_inv ui s\<rbrace>
+   invoke_untyped ui
+   \<lbrace>\<lambda>_ s :: 'state_ext state. pred_tcb_at proj Q t s\<rbrace>"
```
Same with the `P` in this lemma.
